### PR TITLE
feat(windows): listen for DNS change events

### PR DESCRIPTION
**Pull request by ReactorScram**
```[tasklist]
- [x] Ensure whatever method we use to read the DNS servers actually works here, and doesn't have a strange memory ordering problem. If it does, read them from the registry by hand.
- [x] Graceful exit
- [x] Error handling
- [x] Clean it up and integrate it with the Tauri client
- [x] Replace `Notify` with channel of size one
- [x] Remove `Arc`
- [x] Replace `close` with panics
- [x] Remove `Pin`
- [x] Unit-test what happens if we register with RegNotify, close the handle, then modify our key
- [x] Merge with main and call `set_dns`
```

### CODE REVIEW

File: rust/gui-client/src-tauri/Cargo.toml
Comment (ReactorScram) [inline rust/gui-client/src-tauri/Cargo.toml]: I wanted graceful Ctrl+C for the debug subcommand. It's not used in the GUI itself, although there's no reason we couldn't.

Comment (jamilbk) [reply rust/gui-client/src-tauri/Cargo.toml]: I think this is also why the Gateway doesn't respond to Ctrl-C? I think that causes the docker container to delay restart/shutdown

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/gui.rs]: AFAIK `reconnect()` is for network changes and `set_dns(dns_servers: Vec<IpAddr>)` is for DNS. At least that [was my assumption here](https://github.com/firezone/firezone/blob/67c828f013266e85833e8d2b2ffa07eca7f34799/swift/apple/FirezoneNetworkExtension/Adapter.swift#L284).

```suggestion
                    tracing::debug!("New DNS resolvers, calling `Session::set_dns`");
                    // TODO: Pass DNS resolvers to connlib here
                    session.connlib.set_dns([]);
```

Granted on some platforms they'll always be called at the same time (macOS, Android?)

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs:539]: 😆

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/gui.rs]: 👍

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs:1]: Just so I'm following:

- We can't grab DNS-only updates from the Registry
- So we subscribe to whole-interface updates
- When we receive one, we fetch dns_servers again and send them to `Session.set_dns`

Is that about right?

Is it redundant to have the other check then that triggers only `reconnect()`? I.e. can this notify mechanism just trigger the `reconnect()` and `set_dns()` at the same time? Would that be any cleaner?

**Review by jamilbk — COMMENTED** [review]
Nice work! Just some clarifying questions.

Hopefully this is the deepest integration we'll need with the Registry for a while...

File: rust/gui-client/src-tauri/Cargo.toml
Comment (ReactorScram) [reply rust/gui-client/src-tauri/Cargo.toml]: I had thought it was because the `init` flag wasn't set in our Dockerfile, but I can't remember if that fixed it. It seemed to help for the Client.

I've definitely seen it before on other Docker containers, you have to catch Ctrl+C and also tell Docker to forward it.

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs]: You're right, I just thought `set_dns` wasn't merged yet. I'll check again

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs:1]: I'll try it again and document it, but I think this registry key does not change when we gain or lose Internet, because that's an ephemeral thing whereas Windows may store DNS in the registry so it can outlive reboots? 🤔 

And yeah I subscribed to almost anything that changes in the registry relating to any IPv4 or IPv6 interface. Being more specific would add a lot of little edge cases, worst-case resource leaks if I didn't unsubscribe correctly when a network interface was removed. This way it's exactly 2 keys, and I keep them even if Firezone signs out and back in.

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (jamilbk) [reply rust/gui-client/src-tauri/src/client/gui.rs]: Ah, I'm just stubbing it out. I was thinking it was going to be added for this refactor effort

Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs]: Yeah that'll get added in #4163 https://github.com/firezone/firezone/pull/4163/files#diff-0b4003993c22f01feddc49f811f4fde0f8e4111b3c655a6104ee0564efbd5d38R84

Not sure if it makes more sense to keep `reconnect` here or get rid of it and just leave the comments

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs:1]: Hm if I turn Wi-Fi off, wait a bit, and turn it back on, the DNS listener catches when I turn it on (presumably because DHCP kicks in) but not when it goes off.

So if a "Lost Internet" is useful to connlib at all, let's keep both for now.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: Why not remove the `Arc`?

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: We should inline this in `drop`

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/gui.rs]: There's no need to `close` explicitly since it's in `drop`

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: A using a tuple is more standard in rust

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: Why have `Notify` pinned?

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: The callback can decrement the `Arc` using `Arc::from_raw`

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: This should be an `unsafe` function

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: [it's not](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) we should use `mpsc::channel` with buffer size 1 if we want cancellation safety.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: Does these need to be an `Option`? If we don't expose `close` we expect to always have them.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs:548]: just to make sure I understand this `RegisterWaitForSingleObject` will signal `wait_handle`, every time `event` is signaled, even when the event that `event` is listening for changes?

**Review by conectado — COMMENTED** [review]
Looks good! Just left a few comment but nothing big

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I wanted it separate so I can handle errors if closing fails

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: Because of the constraint mentioned in that comment?
🤔 Since it's all `&mut self`, what if I enforce it at runtime with a flag `callback_armed`?

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: Darn. I thought "lose your place in the queue" would be irrelevant if there's only one waiter.
Done

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs:548]: `wait_handle` doesn't signal, it's just a handle we use to un-register the callback when dropping. I added comments here just now: https://github.com/firezone/firezone/pull/4198/files#diff-f8b44e003aaf67d1e8ce86b898c00ae49ebb1ba2ac6d5ae40ec47fdcf64e996dR505-R515

It does fire the callback every time. We only need to call `RegisterWaitForSingleObject` once, but we have to call `RegNotifyChangeKeyValue` repeatedly.

What do you mean, "the event that `event` is listening for changes"?

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: But when `drop` is called, we'll cancel the callback, which will forget the callback's cloned `Arc`, if the callback has not fired. We could check the `Receiver` to see whether we need to force the Arc to drop, but I thought it's simpler to imagine that the callback is taking a const borrow from `Listener`, and `Drop` enforces the lifetimes, and not that they're actually sharing ownership.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I added a comment to explain it, since it's safety-related and I should have it written down: https://github.com/firezone/firezone/pull/4198/files#diff-f8b44e003aaf67d1e8ce86b898c00ae49ebb1ba2ac6d5ae40ec47fdcf64e996dR504-R507

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: It gets shared with another thread. But I have to replace it with an `mpsc` channel anyway per #4226, and when that happens I can remove the `Arc` because it'll be two separate tx and rx pieces.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I think all I do with the errors is log a `tracing::error!` so I could inline that, too. But this felt more right in terms of, "What if someone was actually using this API and it wasn't just tightly integrated with the Windows client"

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: #4227 removes that `Arc`

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: having to explicitly `close` goes against RAII.

I think in this case would even make sense to panic in case of errors.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I mean, it's okay to mark this as unsafe, specially since we use it only internally, the methods using it can make sure that they are enforcing the constaints.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I don't think this needs to be `Pin`, If `Listener` moves, the location pointed to by the `box` wouldn't change since it's heap-allocate, that's why `Box` is `unpin` and why you can still an `&mut T` even when pinned.

Comment (conectado) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: If this wasn't boxed, `Listener` would need to be `!unpin` but I think that'd be harder to implement without much gain

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: this applies here too https://github.com/firezone/firezone/pull/4198#discussion_r1533135656

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (conectado) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: but if we use the `arc` we don't need to unregister the callback, right?

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I don't think it's memory-unsafe, I was thinking of this sentence from the docs https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regnotifychangekeyvalue#remarks

> Each time a process calls RegNotifyChangeKeyValue with the same set of parameters, it establishes another wait operation, creating a resource leak. Therefore, check that you are not calling RegNotifyChangeKeyValue with the same parameters until the previous wait operation has completed.

I might write a unit test for some of the registry stuff, there's no equivalent `Unregister` function, there's just closing the event handle. Even though it's only closed at the end of the process, it would be nice to know that editing the key won't break anything.

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs]: I don't understand. This one is the long-lived callback registered by `RegisterWaitForSingleObject` and unregistered in `Drop` by `UnregisterWaitEx`. If I didn't unregister the callback, it would leak the `Arc` and 1 or 2 Win32 handles.

**Review by conectado — APPROVED** [review]
I've tested that network switching works with this branch, I'd be okay to merge :)

**Comment by ReactorScram** [comment]
@jamilbk Is this really not GA for Windows?

**Comment by ReactorScram** [comment]
Made this GA again since #4262 is GA and is blocked on Windows `set_dns`

**Comment by jamilbk** [comment]
> Made this GA again since https://github.com/firezone/firezone/issues/4262 is GA and is blocked on Windows set_dns

We don't usually set milestones or projects on PRs because it just makes the board too busy. Curious if you're finding it useful though?

PRs will get auto-labeled by CI based on PR title so the changelog is nice, so it's not necessary to label them either, but labeling them doesn't hurt I suppose.

**Comment by ReactorScram** [comment]
@jamilbk Yeah I could use something to prioritize, since I accidentally ended up with IPC stuff that's post-GA, like this: https://github.com/firezone/firezone/pull/4277

If there's a different way I can do it without a milestone and without closing the PR (I'm worried I'd lose track of the branch and not be able to find the PR again when I'm ready to resume it), something else could work.

Edit: Maybe editing the message at the top is enough, usually I put "paused" or "waiting on" or something already.

So normally I try to finish any work-in-progress, starting with things closest to merging, and I want to remember that this PR needs my attention and the IPC PR doesn't.

In general, having PRs and issues separate is a little tough - If I put a tasklist on an issue, I may not remember to check it before merging a PR. But I can't open a PR until the issue is investigated to a certain point, so there's a period where there isn't a good place to "land" more information about an issue.

But that's how all Git forges work, I can't really think of a solution unless Github did a huge change to allow shared tasklists between issues and their PRs. I just like having those little micro-tasks that aren't big enough for a full issue, so that I know I won't forget anything.

**Comment by jamilbk** [comment]
@ReactorScram No worries, I won't change PR milestones then

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (thomaseizinger) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs:447]: Out of curiosity, why make a new runtime here and not just make it an async function?

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (thomaseizinger) [inline rust/gui-client/src-tauri/src/client/network_changes/windows.rs:456]: Personally, I am not touching `tokio::select` with a 10ft pole because it has so many "interesting" features. If you only have two futures, it would suggest using `futures::future::select` instead which is trivial to understand when you look at the implementation.

If it is possible to remove the runtime above, you can even remove the `ctrl_c` listener here and just make this fn a loop of listening to be notified and compose it with `ctrl_c` at the outer layer of your app. I find listening to `ctrl_c` in `main` less surprising than several layers deep here.

YMMV :)

File: rust/gui-client/src-tauri/src/client/network_changes/windows.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs:447]: For the debug commands I don't want to start a runtime until I know which command it'll be, in case one of them needs something odd.

It's pretty rare that it matters, there's only a couple things (COM and Tauri) that could really care what thread they're on or what's happening.

Comment (thomaseizinger) [reply rust/gui-client/src-tauri/src/client/network_changes/windows.rs:447]: I see! I should have checked the code in more detailed, I was assuming that we use `async fn main`!

**Pull request merged**
